File: src/content/docs/topics/cs/coding-problems/advanced-graphs/1584-min-cost-to-connect-all-points-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function minCostConnectPoints(points: number[][]): number {
    const n = points.length;
    const edges: [number, number, number][] = [];
    for (let i = 0; i < n; i++)
        for (let j = i + 1; j < n; j++) {
            const d = Math.abs(points[i][0] - points[j][0]) + Math.abs(points[i][1] - points[j][1]);
            edges.push([d, i, j]);
        }
    edges.sort((a, b) => a[0] - b[0]);

    const parent = Array.from({ length: n }, (_, i) => i);
    function find(x: number): number {
        while (parent[x] !== x) { parent[x] = parent[parent[x]]; x = parent[x]; }
        return x;
    }

    let total = 0, added = 0;
    for (const [d, i, j] of edges) {
        const ri = find(i), rj = find(j);
        if (ri !== rj) {
            parent[ri] = rj;
            total += d;
            if (++added === n - 1) break;
        }
    }
    return total;
}

assert(minCostConnectPoints([[0,0],[2,2],[3,10],[5,2],[7,0]]) === 20);
assert(minCostConnectPoints([[3,12],[-2,5],[-4,1]]) === 18);
assert(minCostConnectPoints([[0,0]]) === 0);
assert(minCostConnectPoints([[0,0],[1,1]]) === 2);
assert(minCostConnectPoints([[0,0],[1,0],[2,0],[3,0]]) === 3);
console.log('all tests pass');

File: src/content/docs/topics/cs/coding-problems/advanced-graphs/1584-min-cost-to-connect-all-points-approach2.py
def min_cost_connect_points(points: list[list[int]]) -> int:
    n = len(points)
    edges = []
    for i in range(n):
        for j in range(i + 1, n):
            d = abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1])
            edges.append((d, i, j))
    edges.sort()

    parent = list(range(n))

    def find(x: int) -> int:
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    total = 0
    edges_added = 0
    for d, i, j in edges:
        ri, rj = find(i), find(j)
        if ri != rj:
            parent[ri] = rj
            total += d
            edges_added += 1
            if edges_added == n - 1:
                break
    return total

assert min_cost_connect_points([[0,0],[2,2],[3,10],[5,2],[7,0]]) == 20
assert min_cost_connect_points([[3,12],[-2,5],[-4,1]]) == 18
assert min_cost_connect_points([[0,0]]) == 0
assert min_cost_connect_points([[0,0],[1,1]]) == 2
assert min_cost_connect_points([[0,0],[1,0],[2,0],[3,0]]) == 3
print("all tests pass")
